applications/serial_lte_modem/src/slm_cmux.c

```diff
@@ -0,0 +1,187 @@
+		const int recv_len = modem_pipe_receive(pipe, recv_buf, sizeof(recv_buf));
+
+		if (recv_len > 0) {
+			LOG_DBG("Received %u bytes on DLCI %u.", recv_len, dlci->address);
+			modem_pipe_transmit(pipe, recv_buf, recv_len);
+		} else {
+			LOG_ERR("Failed to receive data from DLCI %u (%d).",
+				dlci->address, recv_len);
+		}
+		break;
+	}
+}
+
+static void cmux_event_handler(struct modem_cmux *, enum modem_cmux_event event, void *)
+{
+	assert(event == MODEM_CMUX_EVENT_CONNECTED || event == MODEM_CMUX_EVENT_DISCONNECTED);
+	LOG_INF("CMUX %sconnected.", (event == MODEM_CMUX_EVENT_CONNECTED) ? "" : "dis");
+}
+
+static void init_dlci(size_t address, uint16_t recv_buf_size,
+		      uint8_t recv_buf[static recv_buf_size])
+{
+	assert(ARRAY_SIZE(cmux.dlcis) >= address);
+
+	struct cmux_dlci *const dlci = &cmux.dlcis[address - 1];
+	const struct modem_cmux_dlci_config dlci_config = {
+		.dlci_address = address,
+		.receive_buf = recv_buf,
+		.receive_buf_size = recv_buf_size
+	};
+
+	dlci->pipe = modem_cmux_dlci_init(&cmux.instance, &dlci->instance, &dlci_config);
+	dlci->recv_buf = recv_buf;
+	dlci->recv_buf_size = recv_buf_size;
+	dlci->address = address;
+
+	modem_pipe_attach(dlci->pipe, dlci_pipe_event_handler, dlci);
+}
+
+static int cmux_start(void)
+{
+	int ret;
+
+	/* Disable SLM UART so that CMUX can take over it. */
+	ret = slm_uart_rx_disable();
+	if (ret) {
+		return ret;
+	}
+
+	{
+		const struct modem_backend_uart_config uart_backend_config = {
+			.uart = DEVICE_DT_GET(DT_CHOSEN(ncs_slm_uart)),
+			.receive_buf = cmux.uart_backend_receive_buf,
+			.receive_buf_size = sizeof(cmux.uart_backend_receive_buf),
+			.transmit_buf = cmux.uart_backend_transmit_buf,
+			.transmit_buf_size = sizeof(cmux.uart_backend_transmit_buf),
+		};
+
+		cmux.uart_pipe = modem_backend_uart_init(&cmux.uart_backend, &uart_backend_config);
+		if (!cmux.uart_pipe) {
+			return -ENODEV;
+		}
+	}
+	{
+		const struct modem_cmux_config cmux_config = {
+			.callback = cmux_event_handler,
+			.receive_buf = cmux.cmux_receive_buf,
+			.receive_buf_size = sizeof(cmux.cmux_receive_buf),
+			.transmit_buf = cmux.cmux_transmit_buf,
+			.transmit_buf_size = sizeof(cmux.cmux_transmit_buf),
+		};
+
+		modem_cmux_init(&cmux.instance, &cmux_config);
+	}
+	{
+		static uint8_t dlci1_recv_buf[128];
+		static uint8_t dlci2_recv_buf[256];
+
+		init_dlci(1, sizeof(dlci1_recv_buf), dlci1_recv_buf);
+		init_dlci(2, sizeof(dlci2_recv_buf), dlci2_recv_buf);
+	}
+
+	ret = modem_cmux_attach(&cmux.instance, cmux.uart_pipe);
+	if (ret) {
+		return ret;
+	}
+
+	ret = modem_pipe_open(cmux.uart_pipe);
+	return ret;
+}
+
+static void cmux_starter(struct k_work *)
+{
+	const int ret = cmux_start();
+
+	if (ret) {
+		LOG_ERR("Failed to start CMUX (%d).", ret);
+	} else {
+		LOG_INF("CMUX started.");
+	}
+}
+
+/* Handles AT#CMUX commands. */
+int handle_at_cmux(enum at_cmd_type cmd_type)
+{
+	static struct k_work_delayable cmux_start_work;
+	int ret;
+	unsigned int op;
+	enum {
+		OP_START,
```

Comment:
Did I understand correctly that in future you should be able to still send the STOP AT-command via CMUX:ed UART? Or how is this going to be stopped?

Reply:
Possibly. I am a little bit confused regarding this because the 3GPP spec for the `+CMUX` command says
> This command is used to enable/**disable** the 3GPP TS 27.010 [45] multiplexing protocol control channel.

But it doesn't clearly define how to stop CMUX, and even states
> If a +CMUX command is issued whilst in any multiplexer mode then that +CMUX command shall be ignored and the MT/TA shall return a +CME ERROR: <err> response.

Though we may want to be able to stop CMUX regardless of the (admittedly confusing) official spec.
Currently the only way is via an `AT#XRESET`.